src/mscorlib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -777,6 +775,66 @@ public int EnsureCapacity(int capacity)
             return newSize;
         }
 
+        /// <summary>
+        /// Sets the capacity of this dictionary to hold up at least all existing entries
+        /// 
+        /// This method can be used to minimize the memory overhead 
+        /// once it is known that no new elements will be added. 
+        /// 
+        /// To completely clear a dictionary and release all memory 
+        /// referenced by the dictionary, execute the following statements:
+        /// 
+        /// dictionary.Clear();
+        /// dictionary.TrimExcess();
+        /// </summary>
+        public void TrimExcess()
+        {
+            TrimExcess(HashHelpers.GetPrime(Count));
+        }
+
+        /// <summary>
+        /// Sets the capacity of this dictionary to hold up at least 'capacity' entries without any further expansion of its backing storage
+        /// </summary>
+        public void TrimExcess(int capacity)
+        {
+            if (capacity < Count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.capacity);
+            }
+
+            if (_entries != null && capacity < _entries.Length)
+            {
+                int[] buckets = new int[capacity];
+                for (int i = 0; i < buckets.Length; i++)
+                {
+                    buckets[i] = -1;
+                }
+                int count = _count;
+                Entry[] entries = new Entry[capacity];
+                int k = 0;
+                for (int i = 0; i < count; i++)
+                {
+                    if (_entries[i].hashCode >= 0)
+                    {
+                        ref Entry entry = ref _entries[i];
+                        int bucket = entry.hashCode % capacity;
+                        entry.next = buckets[bucket];
+                        buckets[bucket] = i;
+                        entries[k] = entry;
+                        k++;
+                    }
+                }
+                
+                _freeList = -1;
+                _freeCount = 0;
+                _version++;
+                _count = k;
```

Comment:
`_count` is already correct isn't it? no need to set it. you can Debug.Assert _count == k if you want.

Reply:
Since ``` Count = _count - _freeCount``` and in TrimExcess we are clearing _freeCount then _count does get changed

Reply:
Please note, when we call Remove on Dictionary _count remains the same value but _freeCount gets incremented. When TrimExcess gets called and _freeCount would get cleared to zero.

Reply:
Ah yes.